src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -53,6 +53,57 @@ final class HashMap[K, +V] private[immutable] (private[immutable] val rootNode:
 
   override def isEmpty: Boolean = rootNode.size == 0
 
+  override def keySet: Set[K] = if (size == 0) Set.empty else new ImmutableKeySet
+  def oldKeySet: Set[K] = super.keySet
+
+  protected class ImmutableKeySet extends super.ImmutableKeySet {
+
+    private[this] def newKeySetOrThis(newHashMap: HashMap[K, V]) =
+      if (newHashMap eq HashMap.this) this else newHashMap.keySet
+
+    private[this] def toHashSet: HashSet[K] = {
+      def keySetNode(mapNode: MapNode[K, V]): SetNode[K] = mapNode match {
+        case bm: BitmapIndexedMapNode[K, V] =>
+          val bmPayloadArity = bm.payloadArity
+          val bmNodeArity = bm.nodeArity
+          val newContentSize = bmPayloadArity + bmNodeArity
+          val newContent = new Array[Any](newContentSize)
+
+          var i = 0
+          while (i < bmPayloadArity) {
+            newContent(i) = bm.getKey(i)
+            i += 1
+          }
+
+          var j = 0
+          while (j < bmNodeArity) {
+            newContent(newContentSize - 1 - j) = keySetNode(bm.getNode(j))
+            j += 1
+          }
+
+          new BitmapIndexedSetNode[K](bm.dataMap, bm.nodeMap, newContent, bm.originalHashes, bm.size, bm.cachedJavaKeySetHashCode)
+        case hc: HashCollisionMapNode[K, V] =>
+          new HashCollisionSetNode[K](hc.originalHash, hc.hash, hc.content.map(_._1))
+      }
+      new HashSet(keySetNode(rootNode))
+    }
+
+    override def incl(elem: K): Set[K] = if (this(elem)) this else toHashSet + elem
+    override def excl(elem: K): Set[K] = newKeySetOrThis(HashMap.this - elem)
+
+    override def removedAll(that: IterableOnce[K]): Set[K] = newKeySetOrThis(HashMap.this.removedAll(that))
+    override def filter(pred: K => Boolean): Set[K] = newKeySetOrThis(HashMap.this.filter(kv => pred(kv._1)))
+    override def filterNot(pred: K => Boolean): Set[K] = newKeySetOrThis(HashMap.this.filterNot(kv => pred(kv._1)))
+    override def concat(that: IterableOnce[K]): Set[K] = if (that.knownSize == 0) this else toHashSet.concat(that)
+    override def diff(that: collection.Set[K]): Set[K] = newKeySetOrThis(HashMap.this.removedAll(that))
+    override def partition(p: K => Boolean): (Set[K], Set[K]) = {
+      val (left, right) = HashMap.this.partition(kv => p(kv._1))
+      (newKeySetOrThis(left), newKeySetOrThis(right))
+    }
```

Comment:
How did you choose this particular set of methods to override? Some are non-obvious to me.  For instance, `partition` basically has to rebuild the underlying maps.  Wouldn't it be better to just let the default set partition do the rebuilding and not carry around all the extra space of maps?

Reply:
I chose this set of methods because these are methods that can (or do) use structural sharing. I believe all these delegating calls to HashMap are already optimized to share structure except for partition, which has been preemptively overridden in order to allow an optimized implementation to be written later without breaking binary compatibility.

Otherwise, in the case of concat, I've overridden it because the existing implementation will build up a new set element-by-element which should be multiple times slower than the `toHashSet` method implemented here.

You're definitely right though that these intuitions need to be justified with benchmarks. Stay tuned for results!

Reply:
Now I have reduced the scope to just `incl` `excl` `filter` `filterNot`, which are all benchmarked.